src/types/iterator.rs

```diff
@@ -0,0 +1,106 @@
+use crate::convert::FromZvalMut;
+use crate::ffi::zend_object_iterator;
+use crate::flags::DataType;
+use crate::types::{ZendLong, Zval};
+use std::convert::TryInto;
+
+pub type ZendIterator = zend_object_iterator;
+
+pub struct Iter<'a> {
+    zi: &'a mut ZendIterator,
+}
+
+impl ZendIterator {
+    pub fn iter(&mut self) -> Iter {
+        self.index = 0;
+        self.rewind();
+
+        Iter { zi: self }
+    }
+
+    pub fn valid(&mut self) -> bool {
+        if let Some(valid) = unsafe { (*self.funcs).valid } {
+            unsafe { valid(&mut *self) != 0 }
+        } else {
+            true
+        }
+    }
+
+    pub fn rewind(&mut self) {
+        if let Some(rewind) = unsafe { (*self.funcs).rewind } {
+            unsafe {
+                rewind(&mut *self);
+            }
+        }
+    }
+
+    pub fn move_forward(&mut self) {
+        if let Some(move_forward) = unsafe { (*self.funcs).move_forward } {
+            unsafe {
+                move_forward(&mut *self);
+            }
+        }
+    }
+
+    pub fn get_current_data<'a>(&mut self) -> Option<&'a Zval> {
+        let get_current_data = unsafe { (*self.funcs).get_current_data }?;
+        let value = unsafe { &*get_current_data(&mut *self) };
+
+        Some(value)
+    }
+
+    pub fn get_current_key(&mut self) -> Option<Zval> {
+        let get_current_key = unsafe { (*self.funcs).get_current_key }?;
+        let mut key = Zval::new();
+        unsafe {
+            get_current_key(&mut *self, &mut key);
+        }
+
+        Some(key)
+    }
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = (u64, Option<String>, &'a Zval);
```

Comment:
Do you think we could turn the (u64, Option<String>) into an ArrayKey enum of u64/String?

Reply:
I keep the same item type as the array iterator (so an extension may use either array and traversable and keep the same api) but if you prefer that i'm fine with it

Reply:
Actually, I think it might be a nice idea to also change the array iterator api...

Reply:
I have changed both api to use an enum for key